crates/ty_server/src/server/main_loop.rs

```diff
@@ -25,19 +26,48 @@ impl Server {
 
             match next_event {
                 Event::Message(msg) => {
-                    if self.connection.handle_shutdown(&msg)? {
-                        break;
-                    }
+                    let client = Client::new(
+                        self.main_loop_sender.clone(),
+                        self.connection.sender.clone(),
+                    );
+
                     let task = match msg {
                         Message::Request(req) => {
                             self.session
                                 .request_queue_mut()
                                 .incoming_mut()
                                 .register(req.id.clone(), req.method.clone());
 
+                            if self.session.is_shutdown_requested() {
+                                tracing::warn!(
+                                    "Received request after server shutdown was requested, discarding"
+                                );
+                                client.respond_err(
+                                    req.id,
+                                    lsp_server::ResponseError {
+                                        code: lsp_server::ErrorCode::InvalidRequest as i32,
+                                        message: "Shutdown already requested.".to_owned(),
+                                        data: None,
+                                    },
+                                )?;
+                                continue;
+                            }
+
                             api::request(req)
                         }
-                        Message::Notification(notification) => api::notification(notification),
+                        Message::Notification(notification) => {
+                            if notification.method == lsp_types::notification::Exit::METHOD {
+                                tracing::debug!("Received exit notification, exiting");
+                                if !self.session.is_shutdown_requested() {
+                                    tracing::warn!(
+                                        "Received exit notification before shutdown request"
+                                    );
+                                }
+                                return Ok(());
+                            }
+
+                            api::notification(notification)
+                        }
 
                         // Handle the response from the client to a server request
                         Message::Response(response) => {
```

Comment:
I'm assuming we don't need to modify anything in this branch mainly because the server would stop handling any request / notification from the client when shutdown has been initiated and so the server wouldn't try to send any request to the client which means there shouldn't be any client responses to handle during the shutdown process.

Reply:
We could add the handling to notifications, because they're explicitly listed in the LSP specification. It's less clear to me if client responses are excluded too and it's quiet possible that the server might send a request from a background task when the shutdown was already initialized.

Reply:
I'll leave it as is. The LSP specification only mentions:

> If a server receives requests after a shutdown request those requests should error with InvalidRequest.

Processing notifications and responses is a bit wastefull but shouldn't harm too much (they are all very short)